packages/plugins/ens/src/index.js

```diff
@@ -361,31 +361,61 @@ class ENS {
   modifyENSArguments(params, callback) {
     const self = this;
 
-    function checkArgs(argus, cb) {
-      async.map(argus, (arg, nextEachCb) => {
-        if (Array.isArray(arg)) {
-          return checkArgs(arg, nextEachCb);
-        }
-
-        if (!self.isENSName(arg)) {
-          return nextEachCb(null, arg);
-        }
-        self.ensResolve(arg,  (err, address) => {
-          if (err) {
-            return nextEachCb(err);
-          }
-          nextEachCb(null, address);
-        });
-      }, cb);
-    }
-
-    checkArgs(params.contract.args, (err, realArgs) => {
+    checkArgs(params.contract.args, (err, args) => {
       if (err) {
         return callback(err);
       }
-      params.contract.args = realArgs;
+      params.contract.args = args;
       callback(null, params);
     });
+
+    function checkArgs(args, done) {
+      if (Array.isArray(args)) {
+        async.map(args, (arg, next) => {
+          if (Array.isArray(arg)) {
+            return checkArgs(arg, next);
+          }
+
+          if (!self.isENSName(arg)) {
+            return next(null, arg);
+          }
+
+          self.ensResolve(arg,  (err, address) => {
+            if (err) {
+              return next(err);
+            }
+            next(null, address);
+          });
+        }, done);
+      } else {
+        let updatedArgs = {};
+        async.each(Object.keys(args), (key, next) => {
+          const arg = args[key];
+          if (Array.isArray(arg)) {
+            return checkArgs(arg, (err, values) => {
+              updatedArgs[key] = values;
+              next(null);
+            });
+          }
+          if (!self.isENSName(arg)) {
+            updatedArgs[key] = arg;
+            return next(null);
+          }
+          self.ensResolve(arg, (err, address) => {
+            if (err) {
+              return next(err);
+            }
+            updatedArgs[key] = address;
+            next(null);
+          });
+        }, err => {
+          if (err) {
+            return done(err);
+          }
+          done(null, updatedArgs);
+        });
+      }
+    }
```

Comment:
This function looks like it can be made more DRY. If anyone has suggestions, please shout.

Notice that the function bodies of the two code paths are not the same, the array case always `next(null, someValue)` while the object case always `next(null)` (basically `async.map` vs `async.each`).

Reply:
This is not tested, but I think it should work:
```
function checkSingleArg(arg, cb) {
  if (Array.isArray(arg)) {
    return checkArgs(arg, cb);
  }

  if (!self.isENSName(arg)) {
    return cb(null, arg);
  }

  self.ensResolve(arg,  (err, address) => {
    if (err) {
      return cb(err);
    }
    cb(null, address);
  });
}

function checkArgs(args, done) {
  if (Array.isArray(args)) {
    async.map(args, (arg, next) => {
      checkSingleArg(arg, next);
    }, done);
  } else {
    let updatedArgs = {};
    async.each(Object.keys(args), (key, next) => {
      const arg = args[key];

      checkSingleArg(arg, (err, result) => {
        if (err) {
          return next(err);
        }
        updatedArgs[key] = result;
        next():
      });
    }, err => {
      if (err) {
        return done(err);
      }
      done(null, updatedArgs);
    });
  }
}
```